### PR TITLE
Editable image with SHA256

### PR DESCRIPTION
**Pull request by mbilschuk**
Create new editable.

We can drag@drop image inside editable.  
Also editable has to print **sha256** fingerprint of source image and show two tumblnails with effects **rotate180** and **greyscale**.

### CODE REVIEW

File: src/Controller/DropImageController.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ */
+
+namespace App\Controller;
+
+use Pimcore\Model\Asset;
+use Pimcore\Model\Exception\NotFoundException;
+use Symfony\Component\HttpFoundation\Request;
+use Pimcore\Model\Asset\Image;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Template;
+
```
Comment (ykasenka) [inline src/Controller/DropImageController.php:23]: Please remove excess empty line

File: src/Controller/DropImageController.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+class DropImageController extends BaseController {
+
+    /**
+     * Return sha256 fingerprint of image
+     *
+     * @param Image $image
+     * @return string
+     */
+    private function getSha256value(Image $image ): string {
```
Comment (ykasenka) [inline src/Controller/DropImageController.php]:
```php
private function XXX(): string
{
...
}

File: src/Controller/DropImageController.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        $fullPath = $_SERVER['DOCUMENT_ROOT'].'/var/assets'.$image->getRealFullPath();
```
Comment (ykasenka) [inline src/Controller/DropImageController.php]: `$fullPath = $_SERVER['DOCUMENT_ROOT'] . '/var/assets' . $image->getRealFullPath();`

Comment (ykasenka) [inline src/Controller/DropImageController.php]: If it possible, please avoid usage of `$_SERVER` variable. I think in pimcore there is a variable for paths with assets

File: src/Controller/DropImageController.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            if (!is_null($asset) && ('image' === $asset->getType())) {
+                return [
+                    'sha256value' => $this->getSha256value($asset),
+                    'image' => $asset,
+                ];
+            }
+        }
+
+        throw new NotFoundException('Not a image or image not found');
+
```
Comment (ykasenka) [inline src/Controller/DropImageController.php]: Please remove excess empty line

File: src/Controller/DropImageController.php [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+     *
+     * @param Request $request
+     *
+     * @return array
+     */
+    public function imageAction(Request $request)
+    {
+        if ('asset' === $request->get('type')) {
+            $asset = Asset::getById($request->get('id'));
+            if (!is_null($asset) && ('image' === $asset->getType())) {
```
Comment (ykasenka) [inline src/Controller/DropImageController.php]:
```php
if ($asset instanceof Image) {
...
}

File: src/Controller/DropImageController.php (hunk 2 above)
Comment (ykasenka) [inline src/Controller/DropImageController.php]: getSha256**V**alue

File: src/Document/Areabrick/Drop.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ *  @license    http://www.pimcore.org/license     GPLv3 and PEL
+ */
+
+namespace App\Document\Areabrick;
+
+use App\Document\Areabrick\AbstractAreabrick;
+
+class Drop extends AbstractAreabrick
+{
+    public function getName()
```
Comment (ykasenka) [inline src/Document/Areabrick/Drop.php:36]: Please add phpdoc
```php
    /**
     * {@inheritdoc}
     */
```

File: templates/areas/drop/view.html.twig [hunk 7]
```diff
@@ -0,0 +1,12 @@
+<section class="area-standard-teaser bg-light text-center">
+    <div class="container">
+        {{
+            pimcore_renderlet('dropImage', {
+                "controller" : "App\\Controller\\DropImageController::imageAction",
+                "title" : "Drag image asset and Drop it here. You will see <strong>sha256</strong> hash and two thumbnails with some effects",
+                "height" : 600
+            })
+        }}
+
```
Comment (ykasenka) [inline templates/areas/drop/view.html.twig]: Please remove excess empty line

File: templates/default/events.html.twig [hunk 8]
```diff
@@ -1,151 +1,151 @@
-{% extends 'layouts/layout.html.twig' %}
```
Comment (ykasenka) [inline templates/default/events.html.twig]: Please revert back all changes and put only you code

File: templates/default/events.html.twig [hunk 9]
```diff
@@ -1,151 +1,151 @@
...
+                                <p v-if="eventDetails.contactEmail" ><span class="font-weight-bold">Email: </span> [[ eventDetails.contactEmail ]]</p>
+                                <p v-if="eventDetails.contactAddress" ><span class="font-weight-bold">Address: </span> [[ eventDetails.contactAddress ]]</p>
+                            </div>
+                        </div>
+                    </div>
+                </div>
+            </section>
+        </div>
+    </div>
 {% endblock %}
```
Comment (ykasenka) [inline templates/default/events.html.twig]: Please add new line at end

File: templates/drop_image/image.html.twig [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+        </td>
+        <td>
+            {{ image.thumbnail('grayscale').getHtml() | raw }}
+        </td>
+    </tr>
+
+</table>
+
+{% endif %}
+
```
Comment (ykasenka) [inline templates/drop_image/image.html.twig]: Please remove excess empty line

File: templates/drop_image/image.html.twig [hunk 11]
```diff
@@ -0,0 +1,27 @@
+{% if image %}
+
```
Comment (ykasenka) [inline templates/drop_image/image.html.twig]: Please remove excess empty line

File: templates/drop_image/image.html.twig [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+    </tr>
+    <tr>
+        <td>
+            {{ image.thumbnail('rotate180').getHtml() | raw }}
+        </td>
+        <td>
+            {{ image.thumbnail('grayscale').getHtml() | raw }}
+        </td>
+    </tr>
+
```
Comment (ykasenka) [inline templates/drop_image/image.html.twig]: Please remove excess empty line

**Review by ykasenka — CHANGES_REQUESTED** [review]

File: src/Document/Areabrick/Drop.php (hunk 6 above)
Comment (ykasenka) [reply src/Document/Areabrick/Drop.php:36]: Also, if it possible, please use translate for name, like `areabrick.drop.name` and add it to translations

Comment (mbilschuk) [reply src/Document/Areabrick/Drop.php:36]: Done

File: src/Controller/DropImageController.php (hunk 1 above)
Comment (mbilschuk) [reply src/Controller/DropImageController.php:23]: Done

File: src/Controller/DropImageController.php (hunk 2 above)
Comment (mbilschuk) [reply src/Controller/DropImageController.php]: Done

File: src/Controller/DropImageController.php (hunk 3 above)
Comment (mbilschuk) [reply src/Controller/DropImageController.php]: Done

File: src/Controller/DropImageController.php (hunk 4 above)
Comment (mbilschuk) [reply src/Controller/DropImageController.php]: Done

File: src/Controller/DropImageController.php (hunk 3 above)
Comment (mbilschuk) [reply src/Controller/DropImageController.php]: Done

File: src/Controller/DropImageController.php (hunk 5 above)
Comment (mbilschuk) [reply src/Controller/DropImageController.php]: Done

File: src/Controller/DropImageController.php (hunk 2 above)
Comment (mbilschuk) [reply src/Controller/DropImageController.php]: Done

File: templates/areas/drop/view.html.twig (hunk 7 above)
Comment (mbilschuk) [reply templates/areas/drop/view.html.twig]: Done

File: templates/drop_image/image.html.twig (hunk 12 above)
Comment (mbilschuk) [reply templates/drop_image/image.html.twig]: Done

File: templates/drop_image/image.html.twig (hunk 10 above)
Comment (mbilschuk) [reply templates/drop_image/image.html.twig]: Done

File: templates/drop_image/image.html.twig (hunk 11 above)
Comment (mbilschuk) [reply templates/drop_image/image.html.twig]: Done

File: src/Controller/DropImageController.php [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+namespace App\Controller;
+
+use Pimcore\Model\Asset;
+use Pimcore\Model\Exception\NotFoundException;
+use Symfony\Component\HttpFoundation\Request;
+use Pimcore\Model\Asset\Image;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Template;
+
+class DropImageController extends BaseController {
```
Comment (ykasenka) [inline src/Controller/DropImageController.php]:
```php
class XXX extends YYY 
{
...
}
```

File: src/Controller/DropImageController.php [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+     * @Template
+     *
+     * @param Request $request
+     * @return array
+     */
+    public function imageAction(Request $request): array
+    {
+        if ('asset' === $request->get('type')) {
+            $asset = Asset::getById($request->get('id'));
+            if ($asset instanceof  Asset\Image) {
```
Comment (ykasenka) [inline src/Controller/DropImageController.php:48]: Please remove excess space betwen instanceof and Asset\Image

File: templates/areas/drop/view.html.twig [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+    <div class="container">
+        {{
+            pimcore_renderlet('dropImage', {
+                "controller" : "App\\Controller\\DropImageController::imageAction",
+                "title" : "Drag image asset and Drop it here. You will see <strong>sha256</strong> hash and two thumbnails with some effects",
+                "height" : 600
+            })
+        }}
+    </div>
+</section>
```
Comment (ykasenka) [inline templates/areas/drop/view.html.twig]: Please add new line at end

File: templates/drop_image/image.html.twig [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+    <tr>
+        <td>
+            {{ image.thumbnail('rotate180').getHtml() | raw }}
+        </td>
+        <td>
+            {{ image.thumbnail('grayscale').getHtml() | raw }}
+        </td>
+    </tr>
+</table>
+{% endif %}
```
Comment (ykasenka) [inline templates/drop_image/image.html.twig]: Please add new line at end

**Review by ykasenka — CHANGES_REQUESTED** [review]

File: src/Controller/DropImageController.php [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ *  @copyright  Copyright (c) Pimcore GmbH (http://www.pimcore.org)
+ *  @license    http://www.pimcore.org/license     GPLv3 and PEL
+ */
+
+namespace App\Controller;
+
+use Pimcore\Model\Asset;
+use Pimcore\Model\Exception\NotFoundException;
+use Symfony\Component\HttpFoundation\Request;
```
Comment (ykasenka) [inline src/Controller/DropImageController.php:20]: Please organize imports

File: src/Controller/DropImageController.php (hunk 14 above)
Comment (mbilschuk) [reply src/Controller/DropImageController.php:48]: Done

File: src/Controller/DropImageController.php (hunk 17 above)
Comment (ykasenka) [reply src/Controller/DropImageController.php:20]: ^

**Pull request merged**